File: built-in plugins dsp/mb-mallet-musicbox.cpp
/**
 * MB Music Box
 * Category : instrument
 * Type     : mallet
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Delicate music box with tinkling mechanical tone
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MALLET_MUSICBOX_H
#define MB_MALLET_MUSICBOX_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMalletMusicbox : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-mallet-musicbox";
    static constexpr const char* PLUGIN_NAME    = "MB Music Box";
    static constexpr const char* PLUGIN_TYPE    = "mallet";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float brightness = 0.85f;  // range [0, 1]
    float mechanical = 0.4f;  // range [0, 1]
    float reverb = 0.3f;  // range [0, 1]
    float volume = 0.7f;  // range [0, 1]
    };

    MbMalletMusicbox() = default;
    ~MbMalletMusicbox() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.brightness = std::clamp(params.brightness, 0f, 1f);
        params.mechanical = std::clamp(params.mechanical, 0f, 1f);
        params.reverb = std::clamp(params.reverb, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Music Box
        return input;
    }
};

#endif // MB_MALLET_MUSICBOX_H

File: built-in plugins dsp/mb-mallet-musicbox.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbMalletMusicboxPlugin: PluginDefinition = {
  id: "mb-mallet-musicbox",
  slug: "mb-mallet-musicbox",
  name: "MB Music Box",
  category: "instrument",
  type: "mallet" as any,
  version: "1.0.0",
  description: "Delicate music box with tinkling mechanical tone",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sine", detune: 0, gain: 0.6 },
    { type: "sine", detune: 2400, gain: 0.25 },
    { type: "sine", detune: 3600, gain: 0.15 },
  ],
  envelope: { attack: 0.001, decay: 0.8, sustain: 0.0, release: 0.4 },
  parameters: [
    {
      id: "brightness",
      name: "Brightness",
      type: "float",
      defaultValue: 0.85,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "mechanical",
      name: "Mechanical",
      type: "float",
      defaultValue: 0.4,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "reverb",
      name: "Reverb",
      type: "float",
      defaultValue: 0.3,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: {
    brightness: 0.85,
    mechanical: 0.4,
    reverb: 0.3,
    volume: 0.7,
  },
};

export default MbMalletMusicboxPlugin;
